#  Python Data Types


## Learning Objectives
By the end of this train, you should be able to:

* Be aware of the different types of primitive data types;
* Understand the basic operation of each data type; and
* Understand how to assign variables to represent data in your code.

## Outline
In this train we will:

* Highlight the use of each data type;
* Implement the use of built-in functions and methods on different data types;
* Learn how to assign a value to a variable; and
* Understand the naming conventions used for variables.

## Introduction

Python is globally a very popular programming language, used across fields such as software development, data science,  and AI. These use cases require data to be stored efficiently, and in a way that means it can be accessed in a timely manner. Python achieves this with its primary data types. These are the basic building blocks essential for data manipulation and consist of simple values of data.

**Python makes use of four primitive variable types**:
* Integers
* Floats
* Strings
* Booleans

Like other programming languages, variables in Python have values and types. Data types dictate what kind of data a variable can store and in some cases, the amount of memory allocated to it. Allocating variable types allows us to make assumptions about the kind of operations that can be applied to a given variable.

## Numbers

Two main data types exist to store numerical data, for example: `5, 6.21, -5000, 0.002, 9877777`. Numbers can be stored as **integers** or **floats**.

### Integers
Integers, or ints, in Python, are positive or negative whole numbers with no fractional components. By using built-in functions in Python, such as  `type`,  we can determine if a number is an integer or float, as well as convert between the two.

### Floats
The float is another type of numerical data type which is used to represent real numbers. A decimal point is used to divide the integer and fractional components of the value.

**Check data type**

In [1]:
type(5)

int

In [2]:
type(4.9)

float

**Integer to Float**

In [3]:
float(5)

5.0

**Float to Integer**

In [4]:
int(4.9)

4

### Mathematical operations with numbers
Let's look at how we perform mathematical operations with numbers in Python. You will notice that when we have a mathematical operation with an `int` and a `float` the result is always a `float`. This may be due to the conversion that Python performs converting the `int` to a `float` in those cases.

The following numerical operators can be used to perform mathematical operations:
* **Addition**: `+`
* **Subtraction**:  `-`
* **Multiplication**: `*`
* **Division**: `/`
* **Round division**: `//`
* **Exponentiation**: `**`

The code below shows how these operators are used:

In [5]:
#Addition
3 + 3

6

In [6]:
#Subtraction
3 - 2.5

0.5

In [7]:
#Multiplication
3 * 3.5

10.5

In [8]:
#Division
3 / 2

1.5

In [9]:
#Round division
6 // 2

3

In [10]:
#Exponentiation
4 ** 2

16

### BODMAS
Python follows the traditional mathematical rules of precedence, by using the basic order of operations when performing a mathematical calculation (i.e. brackets first, then division and multiplication, and only after that comes addition and subtraction).

 **BODMAS** is an abbreviation for the following:
- **B**rackets
- **O**f
- **D**ivision
- **M**ultiplication
- **A**ddition
- **S**ubtraction

In the example below, Python will first calculate the term  `3 * 3 / 3 `  (which is equal to `3`).  Only then will it add the `3` at the start, and then subtract `2`*(1)*. In order to specify the order of operations - just like at school - we need to make use of brackets*(2)*.

In [11]:
3 + 3 * 3 / 3 - 2 #(1)

4.0

In [12]:
(3 + 3) * 3/(3 - 2) #(2)

18.0

### Modulo
Modulo - `%` is the only operator you might not be familiar with from High School. Modulo calculates the remainder when the number on the left is divided by the number on the right of the modulo.

In [13]:
11 % 2

1

`11 % 3` gives 2 as a result.
This is because the number `3` goes into `11` three times (3 x 3 = 9), with 2 left over (11 - 9 = 2).
*** We often use modulo to determine whether a number is even:***

* **Even number**: if you divide a number by 2 and the result is `0`, the number is *even*
* **Odd number**: if you divide a number by 2 and the result is not `0`, the number is *odd*

In [14]:
#Even number
10 % 2

0

In [15]:
#Odd number
11 % 2

1

### Round
**`round(number, ndigits)`**

We also have a built-in Python function to `round` numbers up/down.For example, to round $\ 3.14159265359\ $(`number`) to $2$ (`digits`) decimals as seen in *(1)*. However, rounding in Python can give unexpected results, for example rounding the float $2.675\ $ to $2\ $ digits, would give $2.67 \ $ and not $\ 2.68\ $ as you would expect *(2)*.

In [16]:
round(3.14159265359, 2) #(1)

3.14

In [17]:
round(2.675, 2) #(2)

2.67

This is not an error. It's just a result of the way computers do math.

Check out https://docs.python.org/3.3/tutorial/floatingpoint.html#tut-fp-issues for more.

## Strings

A ``string``  is a sequence of characters. Anything inside quotes is considered a string in Python, and you can use single (' ') or double (" ") quotes around your string.

`"This is considered a string."`

`'This is also considered as a string.'`

This versatility of strings enables the use of additional quotes and apostrophes within your strings (1). There are other ways of using strings one of these ways is by adding the `string`s (2):

In [18]:
'Python is my favourite subject, Oh wait is it even a subject!' #(1)

'Python is my favourite subject, Oh wait is it even a subject!'

In [19]:
'Financial' + ' ' + 'Calculator' #(2)

'Financial Calculator'

### Differentiating between numbers and strings

When dealing with strings in Python we always need to enclose them in quotation marks, in order to clarify that we want Python to see them as strings. This differentiates them from variable names and numbers.

In [20]:
#add numbers
7 + 7 + 7

21

In [21]:
#add strings
'7' + '7' + '7'

'777'

In [22]:
#multiply variables
7 * 3

21

In [23]:
#Here we just add the string to itself 3 times
'7' * 3

'777'

### Applying Methods on Strings

**`len`** is a method that can be used to work out the length of a `string`, by using **`len( <<YOUR OWN STRING HERE>> )`** *(1)*. You can also use the len function to calculate the total length of multiple strings at once in the same cell *(2)*.

**`upper` & `lower`** are methods that be used to convert a string to either `upper` or `lower` case. We can convert a `string` to `upper`case, by using **`<<YOUR OWN STRING HERE>>.upper()`** *(3)*, or we can convert to `lower` case by using **`<<YOUR OWN STRING HERE>>.lower()`** *(4)*.  

In [24]:
len('Financial') #(1)

9

In [25]:
len('The' + 'driver' + 'has' + 'arrived') #(2)

19

In [26]:
"Calculator".upper() #(3)

'CALCULATOR'

In [27]:
"CALCULATOR".lower() #(4)

'calculator'

### Booleans

In this section we cover **Booleans**, which are built-in data types capable of taking on the values: `True` and `False`. This data type is integral in conditional and comparison expressions and is interchangeable with the integers 1 and 0. Booleans control the flow of a program and are used to make comparisons, showing that despite the fancy name (nicknamed `bool` in Python) they play an integral role.

Boolean data types can be used with `comparison` and `logical` operators:

**Comparison operators**

**==** : Equal to

**!=** : Not Equal to

**<**  : Less than

**>**  : Greater than

**<=** : Less than or equal to

**>=** : Greater than or equal to

**Logical operators**

**and** : True if both are true(`x and y`), this operator is used to check whether both conditions are true

**or**  : True if at least one is true (`x or y`), this operator is used check is either of the conditions is true

**not** : True only if false (`not x`), this operator is used to check for inequality

These Boolean operators will either equate to `True` if the condition is met, or `False` is the condition is not.

#### Comparison operators

Below are some comparison operators that are used to check each condition, run these cells and let's see what you get!

In [28]:
5 < 3

False

In [29]:
5 > 3

True

In [30]:
1 + 1 == 2

True

In [31]:
'apples' == 'oranges'

False

**Logical operators**

Below we will be testing some conditions using logical operators

In [32]:
not 1+1 == 2

False

In [33]:
1+1 != 2

False

In [34]:
'apples' != 'oranges'

True

In [35]:
4 > 2 and 3 > 5

False

In [36]:
True and True

True

In [37]:
True and False

False

In [38]:
False and True

False

In [39]:
False and False

False

In [40]:
4 > 2 or 3 > 5

True

In [41]:
True or True

True

In [42]:
True or False

True

In [43]:
False or True

True

In [44]:
False or False

False

## Variables

We use variables to hold some value (that might change later on). It's often best to imagine variables as "virtual storage boxes" in which we can store stuff. To use our "virtual storage box" we need to give it a name (e.g. 'x') and set it equal to some value (e.g. 253) *(1)*.

When we look below we can see our variable `x` has been assigned to a value. Now we can perform operations with it, e.g. we can multiply it by some other value *(2)*. Variables can store any of the Python data types you have already covered such as; `Integer`, `Float`, `String` and even `Boolean` values. *(3)*. We can even go on and create a variable called 'is_y_greater', that holds the Boolean value to determine whether y is greater than x. *(4)*

#### What is the purpose of using Variables in Python ?

Variables enable us to write flexible programs. Rather than "hard-coding" the data directly into a program, a programmer can use variables to represent the data. Then, when we run the program, the variables are replaced with real data. This makes it possible for the same program to process different sets of data.


In [45]:
x = 253 #(1)

In [46]:
3 * x #(2)

759

In [47]:
#We can find x squared
x ** 2

64009

In [48]:
calculator = "I love Maths "
calculator * 5

'I love Maths I love Maths I love Maths I love Maths I love Maths '

In [49]:
y = 100

In [50]:
is_y_greater = y > x

In [51]:
is_y_greater

False

In [52]:
x + y

353

In [53]:
a = 2
b = 3
c = a**b

In [54]:
c

8

### Overwriting Variables
A variable will always contain the ***last value*** that you assigned to it.

Remember good old **`x`**? (From the very first code line?)

If we decide to change its value from 253 to, say, 5; its as easy as setting **`x`** equal to 5:

In [55]:
x = 5

In [56]:
x

5

In [57]:
x * 3

15

### Name conventions used for Variables
A variable can have a short name (like x and y) or can be defined using a more descriptive name that relates to the variable that you would like to store examples including; `name`, `vehicle` or `total_mass`.

There are some rules to naming variables in Python:

  - A variable name must **start** with a **letter** or the **underscore** character and **cannot a begin with a number**  
  - A variable name can **only contain alpha-numeric characters** and **underscores** (A-z, 0-9, and _ )
  - Variable names are **case-sensitive** (age, Age and AGE are three different variables)
  - Variable names can only be **one word**. Spaces won't do:

In [58]:
vari able = 2

SyntaxError: invalid syntax (4002369568.py, line 1)

In [59]:
3variable = 3

SyntaxError: invalid decimal literal (3890012770.py, line 1)

### Assigning Multiple Variables at Once
In Python, you're even able to assign multiple variables at once. *(1)*. You can also assign the same value to multiple variables in one line. *(2)*

In [67]:
three, five, seven = 3, 5, 7

In [68]:
print(three)
print(five)
print(seven)

3
5
7


In [69]:
a, b, c = "Banana", "Kiwi", "Strawberry"
print(a)
print(b)
print(c)

Banana
Kiwi
Strawberry


In [70]:
a = b =c = "Banana"
print(a)
print(b)
print(c)

Banana
Banana
Banana


In [63]:
x=32
y=130/5
z=6*7

In [64]:
x

32

In [65]:
y

26.0

In [66]:
z

42

## Exercises

### Exercise 1
Express the following math expressions in Python:

**a). $\mathbf{3 \times 3}$**

9

**b). $\mathbf{\frac{144}{12}}$**

12

**c). The remainder after we divide $\mathbf{1565}$ by $\mathbf{3}$**

2

**d). $\mathbf{\sqrt{\pi}}\ \ $  *->    you can use the approximation to $\mathbf{\pi}$ as  $\mathbf{\frac{22}{7}}$***

1.7728105208558367

**e). $\mathbf{\left(\frac{22}{7} + 3.14^{2}\right) \times \frac{3}{0.025}}$**

1560.2948571428572

**f). $\mathbf{\left[\left(\frac{3}{\sqrt{11}}\right)^{2} \times 0.5 \times \frac{7}{9}  \right]^{-1}}$, rounded to 2 decimal places**

3.14

### Exercise 2
Print out your name 500 times by using the `*` in Python:
Make sure that there are spaces between the 500 names.

### Exercise 3
Print out the whole part of $\ \frac{22}{7}$

3

### Exercise 4
Add together: your name in `lower`case and your surname in `upper`case
(Do this by using the `upper` and `lower` functions, and make sure there's a space in the middle)

e.g. `'rendani EMANUEL'`

### Exercise 5
Calculate the `len`gth of your own name (number of characters):

### Exercise 6
Compute the following boolean expressions given:

$x = 2^5$, $y = \frac{130}{5}$ and $z = 6 \times 7$


#### a). y == z

#### b). (y > x) and (z > y)

#### c).  x > (y or z)

#### d). (2x > 3y - 5) or (-y < z - 2x)

### Exercise 7
Which of the following are valid variable names?
#### a). variable 3

#### b). variable_3

#### c). variable3

## Conclusion

We have now learnt what primitive data types exist in python. We have also learnt about built-in functions and methods and how they can be used on the different data types.  You should now have a good understanding of the use of variables when writing code in Python as well as how to assign different data types to variables, and the naming conventions used for variables.

## Appendix

- [Built-in types in Python](https://docs.python.org/3/library/stdtypes.html)

- [Getting started with data](https://runestone.academy/runestone/books/published/pythonds/Introduction/GettingStartedwithData.html#built-in-atomic-data-types)